python/freetensor/core/RPCTools/core.py

```diff
@@ -0,0 +1,104 @@
+import xmlrpc.client as client
+from multiprocessing import Process, Pool
+from xmlrpc.server import SimpleXMLRPCServer
+import socket, sys, time
+from .. import remote_task_scheduler
+
+
+class RPCTool:
+
+    def __init__(self,
+                 remoteTaskScheduler: remote_task_scheduler.RemoteTaskScheduler,
+                 addr="127.0.0.1",
+                 port=8047,
+                 sev_status=3):  #参数是初始化时主服务器的地址
+        """初始化获取主机地址和端口以便分配UUID"""
+        self.UID = 'localhost'  #当无中心服务器分配UID时默认本地运行，分配特殊UID:'localhost
+        self.taskScheduler = remoteTaskScheduler
+        self.serverAddr = self.get_address()
+        self.server = SimpleXMLRPCServer(self.serverAddr, allow_none=True)
+        self.server.register_introspection_functions()
+        self.server.register_multicall_functions()
+        self.server.register_function(self.taskScheduler.remote_task_receive)
+        self.server.register_function(self.taskScheduler.remote_result_receive)
+        self.server.register_function(self.change_status)
+        try:
+            self.serverProcess = Process(target=self.server.serve_forever)
+            self.serverProcess.start()
+            print("RPC Server Started on %s:%d..." %
+                  (self.serverAddr[0], self.serverAddr[1]))
+            self.centerAddr = [addr, port]
+            self.upload(self.centerAddr, sev_status)
+            self.taskScheduler.get_self_uid(self.UID)
+            print("Machine UID:" + self.UID)
+        except KeyboardInterrupt:
+            print("\nKeyboard interrupt received, exiting.")
+
+    def connect(self, addr):
+        """允许失败五次的连接，每次连接之间间隔0.5s"""
+        if "http" not in addr[0]:
+            addr[0] = "http://" + addr[0]
+        for cnt in range(5):
+            try:
+                server = client.ServerProxy(str(addr[0]) + ':' + str(addr[1]))
+                server.check_connection()
+            except:
+                time.sleep(0.1)
+                server = None
+                continue
+            break
+        if server:
+            return server
+        else:
+            raise Exception("Error failed to connect")
+
+    def get_address(self):
+        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        s.bind(('', 0))
+        s.connect(('8.8.8.8', 80))
```

Comment:
What's that?

Reply:
I use it to get the server's IP address and an available port.